src/configure/helper/graphHelper.ts

```diff
@@ -0,0 +1,212 @@
+    private static retryCount = 20;
+    private static retryTimeout = 2 * 1000;
+
+    public static async createSpnAndAssignRole(session: AzureSession, aadAppName: string, scope: string): Promise<AadApplication> {
+        let graphCredentials = await this.getGraphToken(session);
+        let tokenCredentials = new TokenCredentials(graphCredentials.accessToken);
+        let tenantId = session.tenantId;
+        var aadApp: AadApplication;
+
+        return this.createAadApp(tokenCredentials, aadAppName, tenantId)
+        .then((aadApplication) => {
+            aadApp = aadApplication;
+            return this.createSpn(tokenCredentials, aadApp.appId, tenantId);
+        })
+        .then((spn) => {
+            aadApp.objectId = spn.objectId;
+            return this.createRoleAssignment(session.credentials, scope, aadApp.objectId);
+        })
+        .then(() => {
+            return aadApp;
+        })
+        .catch((error) => {
+            let errorMessage = error && error.message;
+            if(!errorMessage && error["odata.error"]) {
+                errorMessage = error["odata.error"]["message"];
+                if(typeof errorMessage === "object") {
+                    errorMessage = errorMessage.value;
+                }
+            }
+            throw new Error(errorMessage);
+        });
+    }
+
+    private static async createAadApp(credentials: TokenCredentials, name: string, tenantId: string): Promise<AadApplication> {
+        let restClient = new RestClient(credentials);
+        let secret = generateRandomPassword(20);
+        let startDate = new Date(Date.now());
+
+        return restClient.sendRequest<any>(<UrlBasedRequestPrepareOptions>{
+            url: `https://graph.windows.net/${tenantId}/applications`,
+            queryParameters: {
+                "api-version": "1.6"
+            },
+            headers: {
+                "Content-Type": "application/json",
+            },
+            method: "POST",
+            body: {
+                "availableToOtherTenants": false,
+                "displayName": name,
+                "homepage": "https://" + name,
+                "identifierUris": [
+                    "https://" + name
+                ],
+                "passwordCredentials": [
+                    {
+                        "startDate": startDate,
+                        "endDate": new Date(startDate.getFullYear() + 1, startDate.getMonth()),
+                        "value": secret
+                    }
+                ]
+            },
+            deserializationMapper: null,
+            serializationMapper: null
+        })
+        .then((data) => {
+            return <AadApplication>{
+                appId: data.appId,
+                secret: secret
+            };
+        });
+    }
+
+    private static createSpn(credentials: TokenCredentials, appId: string, tenantId: string, retries: number = 0): Promise<any> {
+        let restClient = new RestClient(credentials);
```

Comment:
can we reuse rest client

Reply:
yes. I can pass rest client with graph token instead of the token. Keeping both the graph methods as static and don't want to maintain class parameter to avoid handling caching and all.